datashader/transfer_functions/__init__.py

```diff
@@ -574,20 +575,106 @@ def spread(img, px=1, shape='circle', how='over', mask=None, name=None):
         raise ValueError("mask must be a square 2 dimensional ndarray with "
                          "odd dimensions.")
         mask = mask if mask.dtype == 'bool' else mask.astype('bool')
-    kernel = _build_spread_kernel(how)
+
     w = mask.shape[0]
     extra = w // 2
-    M, N = img.shape
-    buf = np.zeros((M + 2*extra, N + 2*extra), dtype='uint32')
-    kernel(img.data, mask, buf)
-    out = buf[extra:-extra, extra:-extra].copy()
-    return Image(out, dims=img.dims, coords=img.coords, name=name)
+    float_type = img.dtype in [np.float32, np.float64]
+
+    if is_image:
+        kernel = _build_spread_kernel(how, is_image)
+    elif float_type:
+        kernel = _build_float_kernel(how, w)
+    else:
+        kernel = _build_int_kernel(how, w, img.dtype == np.uint32)
+
+    if is_image:
+        M, N = img.shape
+        buf = np.zeros((M + 2*extra, N + 2*extra),
+                       dtype='uint32' if is_image else img.dtype)
+        kernel(img.data, mask, buf)
+        out = buf[extra:-extra, extra:-extra].copy()
+    elif float_type:
+        M, N = img.shape
+        buf = np.full((M + 2*extra, N + 2*extra), np.nan, dtype=img.dtype)
```

Comment:
Please pull out the shape calculation to before this `if` statement to avoid repeating it 5 times:

```
shape = (M + 2*extra, N + 2*extra)
```

Also consider whether it's useful to eliminate the other repeated code in these branches to make it more readable. Maybe not, but probably?

Reply:
I agree on pulling out `padded_shape` but I don't think it is clearer to try and pull out much more than that. I find that tackling each dtype separately makes the code easier to think about.

Reply:
Addressed in 9facd5c

Reply:
Sure, just look at it once you've got the shape out and see if anything else jumps out as being better to centralize.  I considered whether the first `if` should be about all those with `len(img.shape)==2`, then only making minor adjustments for type that would then jump out at the reader, but I don't know if that's really going to be easier to read or not without writing it.